msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/DiscussionForumTool.java

```diff
@@ -2224,6 +2236,220 @@ public String processActionToggleDisplayExtendedDescription()
+          }
+        }
+        if (end != null) {  //put the Calendar entry for End in the forum data.
+          if (forumItem instanceof DiscussionTopic) {
+            ((DiscussionTopic) forumItem).setCalendarEndId(end.getId());
+          } else if (forumItem instanceof DiscussionForum) {
+            ((DiscussionForum) forumItem).setCalendarEndId(end.getId());
+          }
+        } else {
+          if (forumItem instanceof DiscussionTopic) {
+            ((DiscussionTopic) forumItem).setCalendarEndId(null);
+          } else if (forumItem instanceof DiscussionForum) {
+            ((DiscussionForum) forumItem).setCalendarEndId(null);
+          }
+        }
+      } else {  // when Send To Calendar is not checked
+        if (calendarBeginId != null){  //don't mess with the Calendar unless there have previously been Calendar events created for this topic
+          try {
+            begin = targetCalendar.getEditEvent(calendarBeginId, CalendarService.EVENT_MODIFY_CALENDAR);
+          } catch (IdUnusedException e) { //if we couldn't get Begin from the calendar, it likely means that the event was deleted from Calendar without consulting Forums, so Forums has an ID saved for a Calendar event that won't exist. We can just create it.
+            begin = targetCalendar.addEvent();
+          }
+          targetCalendar.removeEvent(begin);
+        }
+        if (calendarEndId != null){
+          try {
+            end = targetCalendar.getEditEvent(calendarEndId, CalendarService.EVENT_MODIFY_CALENDAR);
+          } catch (IdUnusedException e) { //this might happen if the event was removed from Calendar without this Forum data being changed, so Forums would have a record for a Calendar event that doesn't exist. We can just create it.
+            end = targetCalendar.addEvent();
+          }
+          targetCalendar.removeEvent(end);
+        }
+        if (forumItem instanceof DiscussionTopic) { //put the null begin/end IDs in the forumItem itself as well, depending on type.
+          ((DiscussionTopic) forumItem).setCalendarBeginId(null);
+          ((DiscussionTopic) forumItem).setCalendarEndId(null);
+        } else if (forumItem instanceof DiscussionForum) {
+          ((DiscussionForum) forumItem).setCalendarBeginId(null);
+          ((DiscussionForum) forumItem).setCalendarEndId(null);
+        }
+      }
+    } catch (IdUnusedException e) { //for getCalendar, if the Calendar doesn't exist.
+      e.printStackTrace();
+    } catch (PermissionException e) { //for addEvent, if the user can't edit a calendar.
+      e.printStackTrace();
+    } catch (InUseException e) {  //for getEditEvent, if it's already being edited.
+      e.printStackTrace();
+    }
+  }
+
+  public Boolean getDoesSiteHaveCalendar(){  //if any Calendar data exists for this site, this method returns True. It's not necessarily based on the literal presence of the Calendar tool in the site, but on Calendar-related data.
```

Comment:
this doesn't seem to be used in this PR

Reply:
It's referenced in the JSPs that revise Forum and Topic settings:
`<h:selectBooleanCheckbox id="sendOpenCloseDateToCalendar"
											 disabled="#{not **ForumTool.doesSiteHaveCalendar**}"
											 value="#{ForumTool.selectedTopic.topic.sendOpenCloseToCalendar}"/>`